src/windows-emulator/windows_emulator.cpp

```diff
@@ -133,15 +137,55 @@ namespace
         return api_set_map_obj;
     }
 
-    emulator_object<API_SET_NAMESPACE> build_api_set_map(x64_emulator& emu, emulator_allocator& allocator)
+    std::vector<uint8_t> obtain_api_set(apiset_location location)
     {
-        // TODO: fix
+        switch (location)
+        {
 #ifdef OS_WINDOWS
-        const auto& orig_api_set_map = *NtCurrentTeb64()->ProcessEnvironmentBlock->ApiSetMap;
-        return clone_api_set_map(emu, allocator, orig_api_set_map);
+        case apiset_location::host: {
+            auto apiSetMap =
+                reinterpret_cast<const API_SET_NAMESPACE*>(NtCurrentTeb64()->ProcessEnvironmentBlock->ApiSetMap);
+            const auto* dataPtr = reinterpret_cast<const uint8_t*>(apiSetMap);
+            std::vector<uint8_t> buffer(dataPtr, dataPtr + apiSetMap->Size);
+            return buffer;
+        }
 #else
-        return clone_api_set_map(emu, allocator, {});
+        case apiset_location::host:
+            throw std::runtime_error("The APISET host location is not supported on this platform");
 #endif
+        case apiset_location::file: {
+            auto apiset = utils::io::read_file("api-set.bin");
+            if (apiset.empty())
+                throw std::runtime_error("Failed to read file api-set.bin");
+            auto buffer = utils::compression::zlib::decompress(apiset);
+            if (buffer.empty())
+                throw std::runtime_error("Failed to decompress API-SET");
+            return buffer;
```

Comment:
maybe extract those 4 lines into a function. they seem to repeat quite a lot?

Reply:
yup, done !